### PR TITLE
HTTP Request Improvement

### LINKED ISSUE #298: Catch and retry subgraph query errors
Cycles and boost will fail with bad requests to the subgraphs, need to have a more robust way to handle. Should catch and retry, and even fail to another subgraph if possible.

```
tests/cycle_utils.py:143: in mock_cycle
    propose_root(
rewards/calc_rewards.py:111: in propose_root
    rewards_data = generate_rewards_in_range(
rewards/calc_rewards.py:199: in generate_rewards_in_range
    all_schedules, setts = fetch_all_schedules(chain, fetch_setts(chain))
rewards/calc_rewards.py:82: in fetch_setts
    setts = list_setts(chain)
subgraph/queries/setts.py:94: in list_setts
    results = client.execute(query)
```

`gql.transport.exceptions.TransportServerError: 502, message='Bad Gateway', url=URL('https://api.thegraph.com/subgraphs/name/axejintao/badger-dao'`

### PR DESCRIPTION
**Pull request by SHAKOTN**
**Done:**
1. Move `http` session to `helpers/` module to avoid circular imports in the future
2. Extend `status_hook`: on `raise_for_status()` send msg to discord about failed request
3. Some logic to avoid situations when trying to get key from the response that returned nothing
4. Add tests to cover all `requests` in `badger-rewards` module
5. Also, added `pytest-mock` library which is standard for pytest mocking and `responses` for mocking requests

Fixes #298 #289

### CODE REVIEW

**Review by Darruma — APPROVED** [review]

File: badger_api/requests.py [hunk 1]
```diff
@@ -46,18 +48,23 @@ def fetch_claimable(page: int, chain: str):
     return response.json()
```
Comment (btcookies) [inline badger_api/requests.py:48]: do we need to if not response here?

File: helpers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from helpers.http_session import http
+
+__all__ = [
```
Comment (btcookies) [inline helpers/__init__.py:3]: do you have a link to this pattern? looks cool never seen before

Comment (SHAKOTN) [reply helpers/__init__.py:3]: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

File: badger_api/requests.py (hunk 1 above)
Comment (SHAKOTN) [reply badger_api/requests.py:50]: Fixed that

**Review by Darruma — APPROVED** [review]

File: tests/test_http_requests_functions.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        responses.GET, "https://api.etherscan.io/api?module=block&action="
+                       "getblocknobytime&timestamp=123&closest=before&apikey=",
+        json={}, status=404
+    )
+    fetch_block_by_timestamp("ethereum", 123)
+
+    # Make sure discord message was sent
+    assert mock_discord.called
+    # Make sure only one message was sent to discord
+    assert mock_discord.call_count == 1
```
Comment (mitche50) [inline tests/test_http_requests_functions.py:34]: Do we want to assert that the response is a correct block here?

**Review by mitche50 — APPROVED** [review]

**Pull request merged**